File: swift/Sources/DeviceAiCore/DeviceAI.swift
/// Primary entry point for the DeviceAI SDK.
///
/// `DeviceAI` is a namespace — it cannot be instantiated.
/// Feature modules extend it with module-specific namespaces via Swift extensions:
///
/// ```swift
/// // DeviceAiLlm adds:   DeviceAI.llm
/// // DeviceAiStt adds:   DeviceAI.stt
/// // DeviceAiTts adds:   DeviceAI.tts
/// ```
///
/// ## Minimal setup (Development — no API key required)
///
/// ```swift
/// // AppDelegate / @main
/// DeviceAI.configure()
///
/// // LLM
/// let session = DeviceAI.llm.chat(modelPath: "/path/to/model.gguf") {
///     $0.systemPrompt = "You are a helpful assistant."
/// }
/// for try await token in session.send("Hello") { print(token, terminator: "") }
///
/// // STT
/// let stt = DeviceAI.stt.session(modelPath: "/path/to/ggml-tiny.bin")
/// let result = try await stt.transcribe(audioPath: "/path/to/audio.wav")
///
/// // TTS
/// let tts = DeviceAI.tts.session(modelPath: "/path/to/voice.onnx", tokensPath: "...")
/// let audio = try await tts.synthesize("Hello world")
/// ```
///
/// ## Kotlin parallel
///
/// ```kotlin
/// // Kotlin — object in :core, extensions in feature modules
/// object DeviceAI { ... }
/// val DeviceAI.llm: LlmModule get() = LlmModule
///
/// // Swift — enum namespace in DeviceAiCore, extensions in feature modules
/// public enum DeviceAI { ... }
/// extension DeviceAI { public static var llm: LLMModule { .shared } }
/// ```
public enum DeviceAI {

    // ── State ─────────────────────────────────────────────────────────────────

    private(set) static var environment: DeviceAiEnvironment = .development
    private(set) static var isConfigured = false

    // ── Initialisation ────────────────────────────────────────────────────────

    /// Configure the SDK at app startup.
    ///
    /// Call once — typically in `AppDelegate.application(_:didFinishLaunchingWithOptions:)`
    /// or the `@main` struct's `init()`. Calling more than once is a no-op after
    /// the first call.
    ///
    /// In `.development` mode no API key is required — all cloud calls are skipped
    /// and models are loaded from explicit local paths you manage.
    ///
    /// - Parameter environment: Target environment. Defaults to `.development`.
    public static func configure(environment: DeviceAiEnvironment = .development) {
        guard !isConfigured else {
            DeviceAiLogger.warning("DeviceAI", "configure() called more than once — ignoring.")
            return
        }
        Self.environment  = environment
        Self.isConfigured = true
        DeviceAiLogger.info("DeviceAI", "Configured — env=\(environment), core=\(SDKVersion.core)")
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    /// Throws `.notInitialised` if `configure()` has not been called.
    /// Feature module sessions call this before touching any native code.
    ///
    /// `package` access — visible to all targets in this SPM package,
    /// but not exported to app code or downstream packages.
    package static func assertConfigured() throws {
        guard isConfigured else { throw DeviceAiError.notInitialised }
    }

    /// Reset SDK state. **For unit tests only** — not part of the public API.
    package static func reset() {
        environment  = .development
        isConfigured = false
    }
}

// ── Environment ───────────────────────────────────────────────────────────────

/// Deployment environment for the SDK.
///
/// Mirrors `Environment` + `apiKey` in kotlin/core — identical semantics.
public enum DeviceAiEnvironment: Equatable, Sendable {
    /// Local development. No API key required. Cloud calls disabled.
    case development

    /// Pre-release QA. Points to staging backend.
    case staging(apiKey: String)

    /// Production. Points to production backend.
    case production(apiKey: String)

    var apiKey: String? {
        switch self {
        case .development:              return nil
        case .staging(let k):           return k
        case .production(let k):        return k
        }
    }
}

File: kotlin/core/src/commonMain/kotlin/dev/deviceai/core/DeviceAI.kt
package dev.deviceai.core

/**
 * Primary entry point for the DeviceAI SDK.
 *
 * Call [initialize] **once** at app startup, then use [llm] and [speech] to run inference.
 *
 * ## Minimal (Development — no backend required)
 * ```kotlin
 * // Android Application.onCreate()
 * DeviceAI.initialize(context)
 *
 * // Then anywhere in your app:
 * val session = DeviceAI.llm.chat("/path/to/model.gguf")
 * session.send("Hello").collect { token -> print(token) }
 * ```
 *
 * ## With cloud (Staging / Production)
 * ```kotlin
 * DeviceAI.initialize(context, apiKey = "dai_live_...") {
 *     environment = Environment.Production
 *     telemetry   = Telemetry.Enabled
 *     wifiOnly    = true
 *     appVersion  = BuildConfig.VERSION_NAME
 *     appAttributes = mapOf("user_tier" to "premium")
 * }
 *
 * // Model path comes from the manifest — no explicit path needed.
 * val session = DeviceAI.llm.chat()
 * ```
 *
 * ## Environments
 * | Environment | API key | Backend | Use for |
 * |---|---|---|---|
 * | [Environment.Development] | not required | none (local) | local dev + unit tests |
 * | [Environment.Staging] | required | staging.api.deviceai.dev | pre-release QA |
 * | [Environment.Production] | required | api.deviceai.dev | release builds |
 */
object DeviceAI {
    internal var cloudConfig: CloudConfig? = null
        private set

    // ── Initialization ────────────────────────────────────────────────────────

    /**
     * Initialize the DeviceAI SDK.
     *
     * **Must be called exactly once** at app startup before using any module.
     * In [Environment.Development] no [apiKey] is required — all cloud calls are
     * skipped and models are loaded from explicit local paths.
     *
     * @param context Android [android.content.Context]. Pass `null` on iOS/JVM.
     * @param apiKey  Your `dai_live_*` key from cloud.deviceai.dev.
     *   Not required in [Environment.Development].
     * @param block   Optional DSL block to configure cloud behaviour, telemetry,
     *   and app metadata. See [CloudConfig.Builder].
     */
    fun initialize(context: Any? = null, apiKey: String? = null, block: CloudConfig.Builder.() -> Unit = {}) {
        val builder = CloudConfig.Builder(apiKey).apply(block)
        val config = builder.build()

        // Configure runtime first — throws IllegalStateException on double-init.
        // Only update cloudConfig after configure() succeeds so state stays consistent.
        DeviceAIRuntime.configure(config.environment)
        cloudConfig = config

        CoreSDKLogger.info(
            "DeviceAI",
            buildString {
                append("Initialized — env=${config.environment}")
                if (config.environment != Environment.Development) {
                    append(", baseUrl=${config.baseUrl}")
                    append(", telemetry=${config.telemetry}")
                }
            },
        )

        if (config.environment == Environment.Development) {
            CoreSDKLogger.debug(
                "DeviceAI",
                "Development mode — cloud calls disabled. " +
                    "Provide model path explicitly: DeviceAI.llm.chat(modelPath)",
            )
            return
        }

        if (apiKey == null) {
            CoreSDKLogger.warn(
                "DeviceAI",
                "apiKey is required for Environment.${config.environment}. " +
                    "Cloud features are disabled. Get your dai_live_* key from cloud.deviceai.dev.",
            )
            return
        }

        // TODO: Phase 2 — start async device registration + manifest sync
        // DeviceRegistration.registerAsync(context, config)
        CoreSDKLogger.debug(
            "DeviceAI",
            "Cloud integration pending Phase 2 — backend not yet connected.",
        )
    }

    // ── Module namespaces ─────────────────────────────────────────────────────
    //
    // LLM:    DeviceAI.llm    — added by kotlin/llm via extension property
    // Speech: DeviceAI.speech — added by kotlin/speech via extension property (Phase 2)

    // ── Observability ─────────────────────────────────────────────────────────

    // TODO: Phase 2 — val status: StateFlow<RuntimeStatus>
    // TODO: Phase 2 — val capabilityTier: CapabilityTier?
    // TODO: Phase 2 — val modelStatus: StateFlow<Map<Module, ModelStatus>>

    /** The environment this instance was initialized with, or null if not yet initialized. */
    val environment: Environment? get() = cloudConfig?.environment

    /** `true` when running in [Environment.Development]. */
    val isDevelopment: Boolean get() = environment == Environment.Development
}
